data_safe_haven/external/api/credentials.py

```diff
@@ -0,0 +1,177 @@
+"""Classes related to Azure credentials"""
+
+import pathlib
+from abc import abstractmethod
+from collections.abc import Sequence
+from datetime import UTC, datetime
+from typing import Any
+
+import jwt
+from azure.core.credentials import AccessToken, TokenCredential
+from azure.identity import (
+    AuthenticationRecord,
+    AzureCliCredential,
+    CredentialUnavailableError,
+    DeviceCodeCredential,
+    TokenCachePersistenceOptions,
+)
+
+from data_safe_haven.exceptions import DataSafeHavenAzureError, DataSafeHavenValueError
+from data_safe_haven.logging import get_logger
+
+
+class DeferredCredential(TokenCredential):
+    """A token credential that wraps and caches other credential classes."""
+
+    token_: AccessToken | None = None
+
+    def __init__(
+        self,
+        scopes: Sequence[str],
+        tenant_id: str | None = None,
+    ) -> None:
+        self.logger = get_logger()
+        self.scopes = scopes
+        self.tenant_id = tenant_id
+
+    @property
+    def token(self) -> str:
+        """Get a token from the credential provider."""
+        return str(self.get_token(*self.scopes, tenant_id=self.tenant_id).token)
+
+    @classmethod
+    def decode_token(cls, auth_token: str) -> dict[str, Any]:
+        try:
+            return dict(
+                jwt.decode(
+                    auth_token,
+                    algorithms=["RS256"],
+                    options={"verify_signature": False},
+                )
+            )
+        except (jwt.exceptions.DecodeError, KeyError) as exc:
+            msg = "Could not interpret input as an Azure authentication token."
+            raise DataSafeHavenValueError(msg) from exc
+
+    @abstractmethod
+    def get_credential(self) -> TokenCredential:
+        """Get a credential provider from the child class."""
+        pass
+
+    def get_token(
+        self,
+        *scopes: str,
+        **kwargs: Any,
+    ) -> AccessToken:
+        # Require at least 10 minutes of remaining validity
+        validity_cutoff = datetime.now(tz=UTC).timestamp() + 10 * 60
+        if not DeferredCredential.token_ or (
+            DeferredCredential.token_.expires_on < validity_cutoff
+        ):
+            # Generate a new token and store it at class-level token
+            DeferredCredential.token_ = self.get_credential().get_token(
+                *scopes, **kwargs
+            )
+        return DeferredCredential.token_
```

Comment:
We could use a [cached property](https://docs.python.org/3/library/functools.html#functools.cached_property) but perhaps this is better.

It will generate a new credential if it needs to?
Is the cut off 10 minutes or 10 hours?

Reply:
The underlying credentials themselves (`AzureCliCredential` and `DeviceCodeCredential` should be able to refresh themselves if needed). `DeviceCodeCredential` actually has token-caching built-in, but `AzureCliCredential` doesn't, and there's a note in the class definition that you should implement this yourself!

Reply:
The cut-off is 10 minutes (see explanatory comment I've added to the code).